epoxy-viewbinder/src/main/java/com/airbnb/epoxy/EpoxyViewBinderVisibilityTracker.kt

```diff
@@ -0,0 +1,256 @@
+package com.airbnb.epoxy
+
+import android.util.Log
+import android.util.SparseArray
+import android.view.View
+import android.view.ViewGroup
+import android.view.ViewTreeObserver
+import androidx.annotation.IdRes
+import androidx.annotation.IntRange
+import androidx.core.view.isVisible
+import androidx.recyclerview.widget.RecyclerView
+import com.airbnb.viewmodeladapter.R
+import java.util.HashMap
+
+/**
+ * A simple way to track visibility events on [EpoxyModel] within an [EpoxyViewBinder].
+ *
+ * [EpoxyViewBinderVisibilityTracker] works with any [View] backed by an [EpoxyModel]. Once attached
+ * the events will be forwarded to the Epoxy model (or to the Epoxy view when using annotations).
+ *
+ * **There are a few exceptions where events are not forwarded:**
+ *  * If a model is replaced with a model of the same class a new impression will not be logged.
+ *  This is due to the view being the same instance and no insight into view holder changes.
+ *  * View binders in scrollable views will only forward the initial visibility state. New events
+ *  will not be emitted on scroll actions. This is due to not knowing when the outer view scrolls.
+ */
+class EpoxyViewBinderVisibilityTracker {
```

Comment:
I suppose it's not feasible or worthwhile to share more of this code with the original visibility tracker?

Reply:
Yeah unfortunately I think they're different enough that it'd be challenging to refactor them to share code. The original one is tightly coupled to working with a recycler view. Some methods are the same/close to the same but I don't think we gain much with the effort to consolidate with compose on the horizon.